File: sdk/python/muzen/remote.py
from __future__ import annotations

import asyncio
import inspect
import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, AsyncIterator, Callable, Dict, List, Optional, Union

from .runner_mapping import (
    _limits_to_runner,
    _session_to_runner,
    _source_to_remote,
)
from .sources import parse_review_source
from .types import (
    ModelProfileInput,
    ProviderProfileInput,
    ReviewArtifact,
    ReviewArtifactExport,
    ReviewArtifactExportOptions,
    ReviewArtifactReadOptions,
    ReviewCancelOptions,
    ReviewEvent,
    ReviewOptions,
    ReviewResult,
    ReviewSessionSnapshot,
    ReviewSourceLike,
)
from .wire_validation import (
    _unwrap_artifact_export,
    _unwrap_model_profile,
    _unwrap_model_profiles,
    _unwrap_optional_review_result,
    _unwrap_provider_profile,
    _unwrap_provider_profiles,
    _unwrap_review_artifact,
    _unwrap_review_events,
    _unwrap_review_snapshot,
)

RemoteTransport = Callable[
    [str, str, Optional[Dict[str, Any]], Dict[str, str]],
    Union[None, Dict[str, Any], List[Any]],
]


class RemoteClient:
    def __init__(
        self,
        *,
        base_url: str,
        token: Optional[str] = None,
        transport: Optional[RemoteTransport] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._token = token
        self._transport = transport

    async def review(
        self,
        source_like: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "RemoteReviewSession":
        source = parse_review_source(source_like)
        payload = {
            "source": _source_to_remote(source),
            "options": _review_options_to_remote(options or ReviewOptions()),
        }
        snapshot = _unwrap_review_snapshot(
            await self._request_json("POST", "/v1/reviews", payload)
        )
        return RemoteReviewSession(self, snapshot)

    async def create_review_session(
        self,
        *,
        source: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "RemoteReviewSession":
        return await self.review(source, options)

    async def resume_review(self, review_id: str) -> "RemoteReviewSession":
        snapshot = _unwrap_review_snapshot(
            await self._request_json("GET", f"/v1/reviews/{_quote(review_id)}")
        )
        return RemoteReviewSession(self, snapshot)

    def workspace(self, workspace_id: str) -> "RemoteWorkspace":
        return RemoteWorkspace(self, workspace_id)

    async def close(self) -> None:
        return None

    async def _request_json(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
    ) -> Any:
        headers = {"Content-Type": "application/json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"
        if self._transport is not None:
            result = self._transport(method, path, body, headers)
            if inspect.isawaitable(result):
                return await result
            return result
        return await asyncio.to_thread(
            _http_json,
            method,
            f"{self._base_url}{path}",
            body,
            headers,
        )


class RemoteWorkspace:
    def __init__(self, client: RemoteClient, workspace_id: str) -> None:
        self._client = client
        self.id = workspace_id
        self.models = RemoteWorkspaceProfileCollection(
            client,
            workspace_id,
            "models",
            _unwrap_model_profile,
            _unwrap_model_profiles,
            _model_profile_input_to_remote,
        )
        self.providers = RemoteWorkspaceProfileCollection(
            client,
            workspace_id,
            "providers",
            _unwrap_provider_profile,
            _unwrap_provider_profiles,
            _provider_profile_input_to_remote,
        )

    async def review(
        self,
        source_like: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "RemoteReviewSession":
        source = parse_review_source(source_like)
        snapshot = _unwrap_review_snapshot(
            await self._client._request_json(
                "POST",
                f"/v1/workspaces/{_quote(self.id)}/reviews",
                {
                    "source": _source_to_remote(source),
                    "options": _review_options_to_remote(options or ReviewOptions()),
                },
            )
        )
        return RemoteReviewSession(self._client, snapshot)


class RemoteWorkspaceProfileCollection:
    def __init__(
        self,
        client: RemoteClient,
        workspace_id: str,
        kind: str,
        unwrap_one: Callable[[Any], Any],
        unwrap_many: Callable[[Any], List[Any]],
        encode_input: Callable[[Any], Dict[str, Any]],
    ) -> None:
        self._client = client
        self._workspace_id = workspace_id
        self._kind = kind
        self._unwrap_one = unwrap_one
        self._unwrap_many = unwrap_many
        self._encode_input = encode_input

    async def set(self, name: str, input: Any) -> Any:
        return self._unwrap_one(
            await self._client._request_json(
                "PUT",
                self._profile_path(name),
                self._encode_input(input),
            )
        )

    async def get(self, name: str) -> Any:
        response = await self._client._request_json("GET", self._profile_path(name))
        if response is None:
            return None
        return self._unwrap_one(response)

    async def list(self) -> List[Any]:
        return self._unwrap_many(
            await self._client._request_json("GET", self._collection_path())
        )

    def _collection_path(self) -> str:
        return f"/v1/workspaces/{_quote(self._workspace_id)}/{self._kind}"

    def _profile_path(self, name: str) -> str:
        return f"{self._collection_path()}/{_quote(name)}"


class RemoteReviewSession:
    def __init__(self, client: RemoteClient, snapshot: ReviewSessionSnapshot) -> None:
        self._client = client
        self.id = snapshot.id
        self.status = snapshot.status
        self.source = snapshot.source
        self._result = snapshot.result

    def subscribe(
        self,
        listener: Callable[[ReviewEvent], None],
        *,
        replay: bool = True,
    ) -> Callable[[], None]:
        cancelled = False

        async def replay_events() -> None:
            async for event in self.events():
                if cancelled:
                    return
                listener(event)

        if replay:
            asyncio.create_task(replay_events())

        def unsubscribe() -> None:
            nonlocal cancelled
            cancelled = True

        return unsubscribe

    async def events(self, *, after: Optional[str] = None) -> AsyncIterator[ReviewEvent]:
        path = f"/v1/reviews/{_quote(self.id)}/events"
        if after:
            path = f"{path}?after={_quote(after)}"
        for event in _unwrap_review_events(
            await self._client._request_json("GET", path)
        ):
            yield event

    async def wait(
        self,
        *,
        timeout: Optional[Union[int, float, str]] = None,
    ) -> ReviewResult:
        timeout_seconds = _parse_timeout_seconds(timeout)
        return await asyncio.wait_for(self._wait_for_result(), timeout=timeout_seconds)

    async def _wait_for_result(self) -> ReviewResult:
        while True:
            result = await self.result()
            if result is not None:
                return result
            await asyncio.sleep(0.25)

    async def result(self) -> Optional[ReviewResult]:
        if self._result is not None:
            return self._result
        self._result = _unwrap_optional_review_result(
            await self._client._request_json(
                "GET", f"/v1/reviews/{_quote(self.id)}/result"
            )
        )
        if self._result is not None:
            self.status = self._result.status
        return self._result

    async def cancel(
        self,
        reason: Optional[Union[str, ReviewCancelOptions]] = None,
    ) -> None:
        cancel_reason = reason.reason if isinstance(reason, ReviewCancelOptions) else reason
        await self._client._request_json(
            "POST",
            f"/v1/reviews/{_quote(self.id)}/cancel",
            {"reason": cancel_reason or "cancelled"},
        )
        self.status = "cancelled"

    async def refresh(self) -> ReviewSessionSnapshot:
        snapshot = _unwrap_review_snapshot(
            await self._client._request_json("GET", f"/v1/reviews/{_quote(self.id)}")
        )
        self.status = snapshot.status
        self._result = snapshot.result
        return snapshot

    async def read_artifact(
        self,
        artifact_id: str,
        options: Optional[ReviewArtifactReadOptions] = None,
    ) -> ReviewArtifact:
        options = options or ReviewArtifactReadOptions()
        return _unwrap_review_artifact(
            await self._client._request_json(
                "GET",
                f"/v1/reviews/{_quote(self.id)}/artifacts/{_quote(artifact_id)}?view={_quote(options.view)}",
            )
        )

    async def export_artifacts(
        self,
        options: Optional[ReviewArtifactExportOptions] = None,
    ) -> ReviewArtifactExport:
        options = options or ReviewArtifactExportOptions()
        return _unwrap_artifact_export(
            await self._client._request_json(
                "POST",
                f"/v1/reviews/{_quote(self.id)}/artifacts/export",
                {
                    "view": options.view,
                    "artifactIds": options.artifact_ids,
                    "maxArtifacts": options.max_artifacts,
                    "maxBytes": options.max_bytes,
                },
            )
        )


def _review_options_to_remote(options: ReviewOptions) -> Dict[str, Any]:
    return {
        "dedupe": options.dedupe,
        "cancelSuperseded": options.cancel_superseded,
        "model": options.model,
        "scope": {
            "files": options.scope_files,
            "include": options.scope_include,
            "exclude": options.scope_exclude,
        },
        "metadata": options.metadata,
        "sessions": [_session_to_runner(session, options.model) for session in options.sessions],
        "limits": _limits_to_runner(options.limits),
    }


def _model_profile_input_to_remote(input: ModelProfileInput) -> Dict[str, Any]:
    return {
        "provider": input.provider,
        "model": input.model,
        "secretRef": input.secret_ref,
        "baseUrl": input.base_url,
        "routing": input.routing,
    }


def _provider_profile_input_to_remote(input: ProviderProfileInput) -> Dict[str, Any]:
    return {
        "provider": input.provider,
        "secretRef": input.secret_ref,
        "baseUrl": input.base_url,
        "routing": input.routing,
    }


def _http_json(
    method: str,
    url: str,
    body: Optional[Dict[str, Any]],
    headers: Dict[str, str],
) -> Any:
    data = json.dumps(body).encode("utf-8") if body is not None else None
    request = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(request) as response:
            if response.status == 204:
                return None
            payload = response.read()
    except urllib.error.HTTPError as error:
        raise RuntimeError(f"Muzen remote request failed: {error.code} {error.reason}") from error
    if not payload:
        return None
    return json.loads(payload.decode("utf-8"))


def _quote(value: str) -> str:
    return urllib.parse.quote(value, safe="")


def _parse_timeout_seconds(timeout: Optional[Union[int, float, str]]) -> Optional[float]:
    if timeout is None:
        return None
    if isinstance(timeout, (int, float)):
        return float(timeout) / 1000
    text = timeout.strip()
    if text.endswith("ms"):
        return float(text[:-2]) / 1000
    if text.endswith("s"):
        return float(text[:-1])
    if text.endswith("m"):
        return float(text[:-1]) * 60
    return float(text) / 1000

File: sdk/typescript/packages/muzen-sdk/src/remote.ts
import { MuzenUnsupportedFeatureError } from "./errors.js";
import {
  isCallbackReviewModelSpec,
  isHostedReviewModelSpec,
} from "./models.js";
import {
  parseTimeoutMs,
  pollUntilResult,
  throwIfAborted,
  withTimeout,
} from "./review-flow.js";
import { parseReviewSource } from "./sources.js";
import { UnsupportedMuzenWorkers } from "./unsupported.js";
import {
  unwrapModelProfile,
  unwrapModelProfiles,
  unwrapOptionalReviewResult,
  unwrapProviderProfile,
  unwrapProviderProfiles,
  unwrapReviewArtifact,
  unwrapReviewArtifactExport,
  unwrapReviewEvents,
  unwrapReviewSnapshot,
} from "./wire-validation.js";
import type {
  CreateMuzenClientOptions,
  ModelProfile,
  ModelProfileInput,
  Muzen,
  MuzenWorkers,
  MuzenWebhookHandler,
  MuzenWebhookProvider,
  MuzenWebhookResponseOptions,
  MuzenWebhooks,
  MuzenWorkspace,
  ProviderProfile,
  ProviderProfileInput,
  ReviewArtifact,
  ReviewArtifactExport,
  ReviewArtifactExportOptions,
  ReviewArtifactReadOptions,
  ReviewCancelOptions,
  ReviewEvent,
  ReviewModelSpec,
  ReviewOptions,
  ReviewResult,
  ReviewSession,
  ReviewSessionSnapshot,
  ReviewSource,
  ReviewSourceLike,
  ReviewStatus,
  SwarmOptions,
  SwarmResult,
  WorkspaceProfileCollection,
} from "./types.js";

export class RemoteMuzen implements Muzen {
  private readonly baseUrl: URL;
  private readonly fetch: typeof globalThis.fetch;
  readonly webhooks: MuzenWebhooks;
  readonly workers: MuzenWorkers;

  constructor(private readonly options: CreateMuzenClientOptions) {
    this.baseUrl = new URL(options.baseUrl);
    this.fetch = options.fetch ?? globalThis.fetch;
    if (!this.fetch) {
      throw new MuzenUnsupportedFeatureError(
        "createMuzenClient({ baseUrl }) requires a fetch implementation",
      );
    }
    this.webhooks = new RemoteMuzenWebhooks(this);
    this.workers = new UnsupportedMuzenWorkers(
      "remote workers are managed by the Muzen service; run workers from a host process created with createMuzen()",
    );
  }

  async review(
    sourceLike: ReviewSourceLike,
    options: ReviewOptions = {},
  ): Promise<ReviewSession> {
    return this.createReviewSession({ source: sourceLike, options });
  }

  workspace(id: string): MuzenWorkspace {
    return new RemoteWorkspace(this, id);
  }

  async createReviewSession(input: {
    source: ReviewSourceLike;
    options?: ReviewOptions;
  }): Promise<ReviewSession> {
    const source = parseReviewSource(input.source);
    const options = input.options ?? {};
    const response = await this.requestJson("/v1/reviews", {
      method: "POST",
      body: {
        source,
        options: remoteReviewOptions(options),
      },
      signal: options.signal,
    });
    const snapshot = unwrapReviewSnapshot(response);
    return new RemoteReviewSession(this, snapshot);
  }

  async runSwarm(_options: SwarmOptions): Promise<SwarmResult> {
    throw new MuzenUnsupportedFeatureError(
      "remote swarm runs are not available yet; use createMuzen() to run swarms against a local muzen-runner",
    );
  }

  async resumeReview(id: string): Promise<ReviewSession> {
    const snapshot = unwrapReviewSnapshot(
      await this.requestJson(`/v1/reviews/${encodeURIComponent(id)}`),
    );
    return new RemoteReviewSession(this, snapshot);
  }

  async close(): Promise<void> {
    // Remote clients do not own a local process.
  }

  async requestJson(
    path: string,
    options: {
      method?: string;
      body?: unknown;
      signal?: AbortSignal;
    } = {},
  ): Promise<unknown> {
    const response = await this.rawRequest(path, {
      method: options.method,
      body:
        options.body === undefined ? undefined : JSON.stringify(options.body),
      signal: options.signal,
      headers: {
        "Content-Type": "application/json",
      },
    });
    if (response.status === 204) {
      return undefined;
    }
    return response.json();
  }

  async rawRequest(path: string, init: RequestInit = {}): Promise<Response> {
    const url = new URL(path, this.baseUrl);
    const headers = new Headers(init.headers);
    if (this.options.token) {
      headers.set("Authorization", `Bearer ${this.options.token}`);
    }
    const response = await this.fetch(url, {
      ...init,
      headers,
    });
    if (!response.ok) {
      throw new Error(
        `Muzen remote request failed: ${response.status} ${response.statusText}`,
      );
    }
    return response;
  }
}

class RemoteMuzenWebhooks implements MuzenWebhooks {
  readonly github: MuzenWebhookHandler;
  readonly gitlab: MuzenWebhookHandler;

  constructor(client: RemoteMuzen) {
    this.github = new RemoteMuzenWebhookHandler(client, "github");
    this.gitlab = new RemoteMuzenWebhookHandler(client, "gitlab");
  }
}

class RemoteMuzenWebhookHandler implements MuzenWebhookHandler {
  constructor(
    private readonly client: RemoteMuzen,
    private readonly provider: MuzenWebhookProvider,
  ) {}

  async response(
    request: Request,
    options: MuzenWebhookResponseOptions = {},
  ): Promise<Response> {
    if (request.method !== "POST") {
      throw new MuzenUnsupportedFeatureError(
        `Muzen ${this.provider} webhook handlers expect POST requests`,
      );
    }
    const body = await request.arrayBuffer();
    const path = options.workspaceId
      ? `/v1/workspaces/${encodeURIComponent(options.workspaceId)}/webhooks/${this.provider}`
      : `/v1/webhooks/${this.provider}`;
    return this.client.rawRequest(path, {
      method: "POST",
      headers: request.headers,
      body,
      signal: options.signal ?? request.signal,
    });
  }
}

class RemoteWorkspace implements MuzenWorkspace {
  readonly models: WorkspaceProfileCollection<ModelProfileInput, ModelProfile>;
  readonly providers: WorkspaceProfileCollection<
    ProviderProfileInput,
    ProviderProfile
  >;

  constructor(
    private readonly client: RemoteMuzen,
    readonly id: string,
  ) {
    this.models = new RemoteWorkspaceProfileCollection(
      client,
      id,
      "models",
      unwrapModelProfile,
      unwrapModelProfiles,
    );
    this.providers = new RemoteWorkspaceProfileCollection(
      client,
      id,
      "providers",
      unwrapProviderProfile,
      unwrapProviderProfiles,
    );
  }

  async review(
    sourceLike: ReviewSourceLike,
    options: ReviewOptions = {},
  ): Promise<ReviewSession> {
    const source = parseReviewSource(sourceLike);
    const response = await this.client.requestJson(
      `/v1/workspaces/${encodeURIComponent(this.id)}/reviews`,
      {
        method: "POST",
        body: {
          source,
          options: remoteReviewOptions(options),
        },
        signal: options.signal,
      },
    );
    return new RemoteReviewSession(this.client, unwrapReviewSnapshot(response));
  }
}

class RemoteWorkspaceProfileCollection<Input, Profile>
  implements WorkspaceProfileCollection<Input, Profile>
{
  constructor(
    private readonly client: RemoteMuzen,
    private readonly workspaceId: string,
    private readonly kind: "models" | "providers",
    private readonly unwrapOne: (value: unknown) => Profile,
    private readonly unwrapMany: (value: unknown) => Profile[],
  ) {}

  async set(name: string, input: Input): Promise<Profile> {
    return this.unwrapOne(
      await this.client.requestJson(this.profilePath(name), {
        method: "PUT",
        body: input,
      }),
    );
  }

  async get(name: string): Promise<Profile | undefined> {
    const response = await this.client.requestJson(this.profilePath(name));
    if (response === undefined || response === null) {
      return undefined;
    }
    return this.unwrapOne(response);
  }

  async list(): Promise<Profile[]> {
    return this.unwrapMany(await this.client.requestJson(this.collectionPath()));
  }

  private collectionPath(): string {
    return `/v1/workspaces/${encodeURIComponent(this.workspaceId)}/${this.kind}`;
  }

  private profilePath(name: string): string {
    return `${this.collectionPath()}/${encodeURIComponent(name)}`;
  }
}

class RemoteReviewSession implements ReviewSession {
  private currentStatus: ReviewStatus;
  private currentResult?: ReviewResult;

  constructor(
    private readonly client: RemoteMuzen,
    snapshot: ReviewSessionSnapshot,
  ) {
    this.id = snapshot.id;
    this.source = snapshot.source;
    this.currentStatus = snapshot.status;
    this.currentResult = snapshot.result;
  }

  readonly id: string;
  readonly source: ReviewSource;

  get status(): ReviewStatus {
    return this.currentStatus;
  }

  subscribe(
    listener: (event: ReviewEvent) => void,
    options: { replay?: boolean } = {},
  ): () => void {
    let cancelled = false;
    if (options.replay ?? true) {
      void (async () => {
        for await (const event of this.events()) {
          if (cancelled) {
            return;
          }
          listener(event);
        }
      })();
    }
    return () => {
      cancelled = true;
    };
  }

  async *events(options: {
    after?: string | null;
    signal?: AbortSignal;
  } = {}): AsyncIterable<ReviewEvent> {
    throwIfAborted(options.signal);
    const search = new URLSearchParams();
    if (options.after) {
      search.set("after", options.after);
    }
    const suffix = search.size > 0 ? `?${search}` : "";
    const response = await this.client.requestJson(
      `/v1/reviews/${encodeURIComponent(this.id)}/events${suffix}`,
      { signal: options.signal },
    );
    const events = unwrapReviewEvents(response);
    for (const event of events) {
      throwIfAborted(options.signal);
      yield event;
    }
  }

  eventsResponse(options: {
    after?: string | null;
    signal?: AbortSignal;
  } = {}): Response {
    const search = new URLSearchParams();
    if (options.after) {
      search.set("after", options.after);
    }
    const suffix = search.size > 0 ? `?${search}` : "";
    const response = this.client.rawRequest(
      `/v1/reviews/${encodeURIComponent(this.id)}/events/stream${suffix}`,
      { signal: options.signal },
    );
    return new Response(
      new ReadableStream({
        async start(controller) {
          try {
            const stream = (await response).body;
            if (!stream) {
              controller.close();
              return;
            }
            const reader = stream.getReader();
            while (true) {
              const next = await reader.read();
              if (next.done) {
                break;
              }
              controller.enqueue(next.value);
            }
            controller.close();
          } catch (error) {
            controller.error(error);
          }
        },
      }),
      {
        headers: {
          "Content-Type": "text/event-stream",
        },
      },
    );
  }

  async wait(options: {
    timeout?: string | number;
    signal?: AbortSignal;
  } = {}): Promise<ReviewResult> {
    throwIfAborted(options.signal);
    const timeoutMs = parseTimeoutMs(options.timeout);
    const resultPromise = pollUntilResult(
      () => this.result(options),
      options.signal,
    );
    if (timeoutMs === undefined) {
      return resultPromise;
    }
    return withTimeout(resultPromise, timeoutMs);
  }

  async result(options: { signal?: AbortSignal } = {}): Promise<ReviewResult | undefined> {
    if (this.currentResult) {
      return this.currentResult;
    }
    const response = await this.client.requestJson(
      `/v1/reviews/${encodeURIComponent(this.id)}/result`,
      { signal: options.signal },
    );
    this.currentResult = unwrapOptionalReviewResult(response);
    if (this.currentResult) {
      this.currentStatus = this.currentResult.status;
    }
    return this.currentResult;
  }

  async readArtifact(
    artifactId: string,
    options: ReviewArtifactReadOptions = {},
  ): Promise<ReviewArtifact> {
    const search = new URLSearchParams({
      view: options.view ?? "redacted",
    });
    const response = await this.client.requestJson(
      `/v1/reviews/${encodeURIComponent(this.id)}/artifacts/${encodeURIComponent(artifactId)}?${search}`,
    );
    return unwrapReviewArtifact(response);
  }

  async exportArtifacts(
    options: ReviewArtifactExportOptions = {},
  ): Promise<ReviewArtifactExport> {
    return unwrapReviewArtifactExport(
      await this.client.requestJson(
        `/v1/reviews/${encodeURIComponent(this.id)}/artifacts/export`,
        {
          method: "POST",
          body: options,
        },
      ),
    );
  }

  async cancel(reason?: string | ReviewCancelOptions): Promise<void> {
    const cancelReason =
      typeof reason === "string" ? reason : reason?.reason ?? "cancelled";
    await this.client.requestJson(
      `/v1/reviews/${encodeURIComponent(this.id)}/cancel`,
      {
        method: "POST",
        body: { reason: cancelReason },
      },
    );
    this.currentStatus = "cancelled";
  }

  async refresh(): Promise<ReviewSessionSnapshot> {
    const snapshot = unwrapReviewSnapshot(
      await this.client.requestJson(`/v1/reviews/${encodeURIComponent(this.id)}`),
    );
    this.currentStatus = snapshot.status;
    this.currentResult = snapshot.result;
    return snapshot;
  }
}

function remoteReviewOptions(options: ReviewOptions): ReviewOptions {
  const {
    hooks: _hooks,
    signal: _signal,
    heartbeat: _heartbeat,
    sourceProvider,
    model,
    tools,
    ...serializable
  } = options;
  return {
    ...serializable,
    model: remoteModel(model),
    sessions: serializable.sessions?.map(remoteSession),
    sourceProvider: sourceProvider?.baseUrl
      ? { baseUrl: sourceProvider.baseUrl }
      : undefined,
    tools: tools?.map(({ handler: _handler, ...tool }) => tool),
  };
}

function remoteSession(
  session: NonNullable<ReviewOptions["sessions"]>[number],
): NonNullable<ReviewOptions["sessions"]>[number] {
  return {
    ...session,
    model: remoteModel(session.model),
  };
}

function remoteModel(
  model: ReviewOptions["model"],
): ReviewModelSpec | undefined {
  if (isHostedReviewModelSpec(model)) {
    return model;
  }
  if (isCallbackReviewModelSpec(model)) {
    return undefined;
  }
  return undefined;
}
